Xmltv Guide data

fuzzard opened the pull request:
As discussed in https://github.com/kodi-pvr/pvr.hdhomerun/issues/57

Any feedback or questions would be appreciated. 

There is still a bit of refinement to be done, particularly around some existing json fields provided by the SiliconDust data to be pulled from the xmltv file. Genres need to be done, Series ID, episode numbers, etc.

As it stands, the selector only accepts a local file, however the function works with a web hosted xml/xml.gz. Will implement an option for URL/local later.

fuzzard commented:
Have implemented genres from xmltv category elements
Added setting for URL or local file for xmltv data

fuzzard commented:
I consider this feature complete for an initial PR. Obviously if any changes are requested, i can get them done as quickly as i can. 
Alternatively if you believe this isnt suitable/wanted, let me know and i can close the PR and leave it at that.

zcsizmadia commented:
Thanks! Let me build the addon from you PR branch and check it out. It will be most likely this weekend or early next week.

fuzzard commented:
Ive added a commit solely focused on the use of rapidxml. There is a build failure that happens when using msvc, and to get around this other addons (iptvsimple from memory) alter the header file to get around the fault (existing function name).
I think the better way to handle this is to patch the file rather than manually edit it. This at least allows someone to see that the rapidxml header is altered from the 1.13 upstream, and if it ever does get updated, we can assess the patch, rather than alter the source.
If you know of a better method, let me know, but ive added a step in the cmake build script to apply the patch before compilation. Im not sure if we could use cmake's externalproject_add, as the headers are not compiled libraries. Ill stick with this method for now unless you have better ideas.

fuzzard commented:
Removed the jsoncpp update from this PR and did a separate PR. I figure thats an easy win that should be done regardless.

fuzzard commented:
Rebased and changed a few cosmetics and removed some unsused variables
another for range loop changeover, and removed use of the String typedef new code in anticipation of a pending PR that removes the typedef from the codebase.

dr88dr88 commented:
Is it not better to develop a generic EPG scrapper infrastructure that can combine EPG from multiple sources via EPG add-ons and so enhance all EPG data for all PVR clients.

fuzzard commented:
That would be a question better directed to one of the kodi dev's. At a guess though, i doubt it would be something they would want/need to implement inside kodi. I believe all pvr addons but iptvsimple have guide data sent from the backend (this addon included). There are plenty of 3rd party scrapers (zap2xml for 1 example).
The only reason i even thought about doing this is the standard guide provided with a hdhomerun only shows 4hrs (though you can extend that to 12 hours as seen in this PR), unless u use the dvr servce and then it provides 24hr i believe (not exactly sure as i do not subscribe to the dvr service). I was intending to look at implementing the recording functionality (see other PR), and obviously with only a 4hr guide, it makes some of the fancier timers available in other addons not possible at all.

dr88dr88 commented:
When will this pull request be ready to be merged. I like to use it in combination with WebGrab+

fuzzard commented:
It's been ready to merge for months. Depends on whether maintainers want to have it or not though

dr88dr88 commented:
Dear add-on maintainer can you please merge this pull request.

dr88dr88 commented:
I have the feeling that the orginal maintainer is not looking into this add-on any more. 
Fuzzard could you not take over as the roll as maintainer of this add-on.

Second question Fuzard do you have a zip file for the add-on for platform linux-i686 so I can test out your changes. I am using Milehouse latest Libreelec release.

fuzzard commented:
I dont consider myself good enough to maintain the project. Ive been very greatful to ksooo for all of his reviews, so the answer there i believe would be a no. I wouldnt trust the quality of the commits to be of the standard Kodi deserves/requires if i was maintainer.

I may be able to create a zip, ill have a look into it. I dont have a linux dev (i use OSX) environment maintained at the moment, but ill do up a VM, and see what i can get for you.

dr88dr88 commented:
@fuzzard a zip would be great. 
@ksooo could you look if this pull request can be merged.

fuzzard commented:
It's not really up ksooo's alley. He's not into the details of other PVR add-ons, and I'm sure he is busy doing important things in core.

dr88dr88 commented:
@fuzzard I moved my setup to Tvheadend to solved my EPG issue, as bonus I also got time shifting and recording. The binary .zip are not needed anny more for me. Thanks for your help.

Rechi commented on CMakeLists.txt:
```diff
@@ -10,30 +10,47 @@ find_package(Kodi REQUIRED)
+  FILE( WRITE ${RAPIDXML_DIR}/${PATCH_BATCH_FILE} "${PATCH_EXECUTABLE} -p0 -i ${RAPIDXML_DIR}/0001-Win-compile.patch" )
+  EXECUTE_PROCESS( COMMAND cmake -E chdir ${RAPIDXML_DIR} ${PATCH_BATCH_FILE} RESULT_VARIABLE RESULT_PATCH )
+  MESSAGE( STATUS "Patch result: ${RESULT_PATCH}, PROF SOURCE DIR: ${PROJECT_SOURCE_DIR}" )
+  FILE( REMOVE ${PATCH_BATCH_FILE} )
```
such hacks aren't allowed here

fuzzard replied:
Out of curiosity, how should it be handled. The context for this is below:
Chose rapidxml as it is in use by another binary pvr addon (the name eludes me at the moment).
When i was looking into dependency updates, i noticed that the binary addon ships a prepatched rapdixml library, with no mention/info about the fact it is patched. This isnt really a problem in my eyes, but as i was adding this i thought it may be best to use direct from upstream, and patch in build process. This gives more transparency that a patch is applied, and only on windows builds, and would allow the removal if upstream ever fixes it.
The patch is only required for windows builds, and there is little/no up stream response, so trying to get it patched upstream isnt happening.

Do i just ship the patched library? Anywhere suitable for notes/info incase someone else walks by to update at a later date to inform them that the library is in fact patched and not direct from upstream?

Rechi replied:
I'll open a PR for the other one soon, which moves rapidxml to depends.

fuzzard replied:
Thanks, saw the iptv.simple PR for the move. I've duplicated that here.

Rechi commented on lib/rapidxml/0001-Win-compile.patch:
```diff
@@ -0,0 +1,43 @@
+--- rapidxml.hpp	2017-09-07 18:25:10.000000000 +0700
```
rapidxml lib has to be moved to depends

fuzzard committed:
implement XMLTV Guide usage

Implements settings to allow usage of xmltv file/url for guide data instead of SD's
Guide. Also implements an extended SD guide (24hrs without DVR subscription)

Rechi commented on FindRapidxml.cmake:
```diff
@@ -0,0 +1,12 @@
+include(FindPackageHandleStandardArgs)
+find_package_handle_standard_args(Rapidxml DEFAULT_MSG RAPIDXML_INCLUDE_DIRS)
+
+mark_as_advanced(RAPIDXML_INCLUDE_DIRS)
```
always add a new line at the end of file

Rechi commented on appveyor.yml:
```diff
@@ -15,6 +15,15 @@ environment:
+  - ps: copy "C:\Program Files\Git\usr\bin\patch.exe" "C:\Program Files\PatchFromGit"
+  - ps: copy "C:\Program Files\Git\usr\bin\msys*.dll" "C:\Program Files\PatchFromGit"
+  - set PATH=C:\Program Files\PatchFromGit;%PATH%
+
```
this isn't needed

Rechi commented on depends/common/rapidxml/CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+cmake_minimum_required(VERSION 3.5)
+ project(rapidxml)
+ install(FILES rapidxml.hpp rapidxml_iterators.hpp rapidxml_print.hpp rapidxml_utils.hpp DESTINATION include/rapidxml)
```
new line

Rechi commented on depends/common/rapidxml/rapidxml.sha256:
```diff
@@ -0,0 +1 @@
+c3f0b886374981bb20fabcf323d755db4be6dba42064599481da64a85f5b3571
```
new line

Rechi commented on depends/common/zlib/zlib.txt:
```diff
@@ -0,0 +1 @@
+zlib http://mirrors.xbmc.org/build-deps/sources/zlib-1.2.11.tar.gz
```
use a newer zlib version, you can take it from https://github.com/xbmc/audiodecoder.2sf/commits/master/depends/common/zlib

Rechi replied:
You are right version is already the latest released one.
You should replace the whole folder (remove CMakeLists.txt and add 01-build-static.patch, zlib.sha256 from the linked addon)

fuzzard replied:
Cheers, Done.

ksooo commented on src/EPG_SD.cpp:
```diff
@@ -0,0 +1,157 @@
+  return 0;
+}
+
+// TODO: i not used
```
if it is not used, remove it. ;-)

ksooo commented on src/EPG_SD.cpp:
```diff
@@ -0,0 +1,157 @@
+
+}
+
+bool CEpg_SD::UpdateBasicGuide(HDHomeRunTuners::Tuner *pTuner, std::string strUrl)
```
`const std::string&` ?

ksooo commented on src/EPG_SD.cpp:
```diff
@@ -0,0 +1,157 @@
+{
+  std::string strJson, jsonReaderError;
+  Json::CharReaderBuilder jsonReaderBuilder;
+  std::unique_ptr<Json::CharReader> const jsonReader(jsonReaderBuilder.newCharReader());
```
declare vars where actually used

ksooo commented on src/EPG_SD.cpp:
```diff
@@ -0,0 +1,157 @@
+    if (jsonReader->parse(strJson.c_str(), strJson.c_str() + strJson.size(), &pTuner->Guide, &jsonReaderError) &&
+        pTuner->Guide.type() == Json::arrayValue)
+    {
+      for (auto& iterGuide : pTuner->Guide)
```
`const auto&` ?

ksooo commented on src/EPG_SD.cpp:
```diff
@@ -0,0 +1,157 @@
+    {
+      for (auto& iterGuide : pTuner->Guide)
+      {
+        Json::Value& jsonGuide = iterGuide["Guide"];
```
`const Json::Value` ?

ksooo commented on src/EPG_SD.h:
```diff
@@ -0,0 +1,51 @@
+ *      Copyright (C) 2015 Zoltan Csizmadia <zcsizmadia@gmail.com>
+ *      https://github.com/zcsizmadia/pvr.hdhomerun
+ *
+ *      Copyright (C) 2011 Pulse-Eight
```
Why Pulse-Eight?

ksooo commented on src/EPG_SD.h:
```diff
@@ -0,0 +1,51 @@
+
+class CEpg_SD : public CEpgBase
+{
+
```
remove this line?

ksooo commented on src/EPG_SD.h:
```diff
@@ -0,0 +1,51 @@
+{
+
+  public:
+    CEpg_SD(){};
```
CEpg_SD() = default;

ksooo commented on src/EPG_SD.h:
```diff
@@ -0,0 +1,51 @@
+
+  public:
+    CEpg_SD(){};
+    virtual ~CEpg_SD(){};
```
virtual ~CEpg_SD() = default;

ksooo commented on src/EPG_SD.h:
```diff
@@ -0,0 +1,51 @@
+    unsigned long long GetEndTime(Json::Value jsonGuide);
+    void InsertGuideData(Json::Value &Guide, Json::Value strInsertdata);
+    bool UpdateAdvancedGuide(HDHomeRunTuners::Tuner *pTuner, std::string strUrl);
+    bool UpdateBasicGuide(HDHomeRunTuners::Tuner *pTuner, std::string strUrl);
```
by-value vs. by-ref, const vs. non-const

ksooo reviewed (commented):
Boah, and my comments only cover a fist chunk of the changes... 

Many similar topics in the rest. Maybe you get the pattern (copright pulse-eight, params by-ref vs. by-value, const vs. non-const, scope of variales, code formatting, ...) and change the code proactively.

fuzzard commented:
I'll look into all those this weekend. I do appreciate it.
As for the copyright stuff, some of the code is just refactored from the current code base. I assumed the copyright to that original code still carries if chunks have been copied unaltered?  I'm really in the dark when it comes to IP and copyright stuff unfortunately.

fuzzard commented:
First round of fixups are all done. the cmake/library stuff requested by rechi. I believe all the const/ref/indentation should be ok for everything now.
I havent touched the copyright stuff, as in the above comment, there is a bunch of code that was just relocated. I've no idea if it was zoltan or pulse8 who created it, so have copied copyright header into the files that have any code from them.
Whats easier from a review viewpoint, fixup commits, or rebase into the original commit?

fuzzard commented on src/EPG_SD.cpp:
```diff
@@ -0,0 +1,157 @@
+  if (pTuner->Guide.type() == Json::arrayValue)
+  {
+    //Loop each channel Guide
+    for (auto& iterGuide : pTuner->Guide)
```
cant be const on this one. Gets a bit muddy, but the iterGuides are all updated through their lifetime.

ksooo commented:
rebase, please.

ksooo commented:
Keeping that copyright stuff as it is now is fine by me. IANAL ;-)

ksooo commented:
Please reopen for Matrix if you feel this is still of interest.
